### PR TITLE
BUGFIX: SD-494 disable saturate other element on hover for combo line line

### PR DESCRIPTION
**Pull request by lanhq**
[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [x] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [x] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [ ] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [x] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [x] Successful `extended test - examples`
- [x] Successful `extended test - storybook`
- [x] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)

# Related PRs

- gdc-analytical-designer: https://github.com/gooddata/gdc-analytical-designer/pull/2290
- gdc-dashboards: https://github.com/gooddata/gdc-dashboards/pull/2024

# Related Jira tasks
- SD-494: https://jira.intgdc.com/browse/SD-494

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3929/ : SUCCESS in 4m 29s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4973/ : SUCCESS in 11s

**Comment by lanhq** [comment]
extended test - storybook

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2149/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1497/ : SUCCESS in 13m 41s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4976/ : SUCCESS in 7s

**Comment by lanhq** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1025.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1498/ : SUCCESS in 12m 30s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4977/ : SUCCESS in 9s

File: src/components/visualizations/chart/highcharts/comboConfiguration.ts [hunk 1]
```diff
@@ -24,9 +24,22 @@ export function getDefaultChartType(config?: IChartConfig) {
 }
 
 export function getComboConfiguration(config?: IChartConfig) {
+    const chartType = getDefaultChartType(config);
+    const series =
+        chartType === LINE
```
Comment (naibinh) [inline src/components/visualizations/chart/highcharts/comboConfiguration.ts]: you can do `isLineChart(chartType)`

File: src/components/visualizations/chart/highcharts/test/comboConfiguration.spec.ts [hunk 2]
```diff
@@ -36,4 +37,26 @@ describe("Combo Configuration", () => {
             expect(getDefaultChartType(config)).toEqual(LINE);
         });
     });
+
+    it("should disable saturate other for combo line_line", () => {
+        const config = {
```
Comment (naibinh) [inline src/components/visualizations/chart/highcharts/test/comboConfiguration.spec.ts]: we should add type to `config`

File: src/components/visualizations/chart/highcharts/test/comboConfiguration.spec.ts [hunk 3]
```diff
@@ -1,6 +1,7 @@
 // (C) 2019 GoodData Corporation
-import { getDefaultChartType } from "../comboConfiguration";
+import { getDefaultChartType, getComboConfiguration } from "../comboConfiguration";
 import { VisualizationTypes } from "../../../../../constants/visualizationTypes";
+import get = require("lodash/get");
```
Comment (naibinh) [inline src/components/visualizations/chart/highcharts/test/comboConfiguration.spec.ts]: move this absolute import to first file please

File: src/components/visualizations/chart/highcharts/test/comboConfiguration.spec.ts [hunk 4]
```diff
@@ -36,4 +37,26 @@ describe("Combo Configuration", () => {
...
+    it("should not disable saturate other elements for another kind of combo", () => {
+        const config = {
+            primaryChartType: COLUMN,
+            secondaryChartType: LINE,
+        };
+        const {
+            plotOptions: { series: series },
+        } = getComboConfiguration(config);
+        expect(get(series, "states.inactive.opacity")).not.toBe(1);
+    });
```
Comment (naibinh) [inline src/components/visualizations/chart/highcharts/test/comboConfiguration.spec.ts]: how about adding more test with empty `config`, cause it is optional

Comment (lanhq) [reply src/components/visualizations/chart/highcharts/test/comboConfiguration.spec.ts]: done

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: src/components/visualizations/chart/highcharts/comboConfiguration.ts [hunk 5]
```diff
@@ -23,23 +24,27 @@ export function getDefaultChartType(config?: IChartConfig) {
     return LINE;
 }
 
+function isLineLineCombo(config?: IChartConfig) {
+    const primaryType = get(config, "primaryChartType", COLUMN);
+    const secondaryType = get(config, "secondaryChartType", LINE);
```
Comment (naibinh) [inline src/components/visualizations/chart/highcharts/comboConfiguration.ts]: we can avoid using get, it returns `any`
```
const primaryType = config.primaryChartType || COLUMN;
const secondaryType = config.secondaryChartType || LINE;
```

File: src/components/visualizations/chart/highcharts/comboConfiguration.ts [hunk 6]
```diff
@@ -23,23 +24,27 @@ export function getDefaultChartType(config?: IChartConfig) {
     return LINE;
 }
 
+function isLineLineCombo(config?: IChartConfig) {
```
Comment (naibinh) [inline src/components/visualizations/chart/highcharts/comboConfiguration.ts]: `boolean` as return type please

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: src/components/visualizations/chart/highcharts/comboConfiguration.ts (hunk 5 above)
Comment (lanhq) [reply src/components/visualizations/chart/highcharts/comboConfiguration.ts]: Use lodash/get here is acceptable. because `config` is optional.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: src/components/visualizations/chart/highcharts/comboConfiguration.ts (hunk 5 above)
Comment (naibinh) [reply src/components/visualizations/chart/highcharts/comboConfiguration.ts]: so could it have default value?

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3936/ : SUCCESS in 17m 56s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4983/ : SUCCESS in 10s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3962/ : FAILURE in 3m 57s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5014/ : SUCCESS in 7s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3964/ : SUCCESS in 8m 31s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5016/ : SUCCESS in 7s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3967/ : SUCCESS in 8m 08s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5021/ : SUCCESS in 6s

**Comment by naibinh** [comment]
LGTM

File: src/components/visualizations/chart/highcharts/comboConfiguration.ts [hunk 7]
```diff
@@ -1,29 +1,47 @@
...
 const { COLUMN, LINE } = VisualizationTypes;
+const DEFAULT_COMBO_CONFIG: IChartConfig = {
+    primaryChartType: COLUMN,
+    secondaryChartType: LINE,
+};
 
-export function getDefaultChartType(config?: IChartConfig) {
-    const primaryType = get(config, "primaryChartType", COLUMN);
-    const secondaryType = get(config, "secondaryChartType", LINE);
+export function getDefaultChartType(config: IChartConfig = DEFAULT_COMBO_CONFIG) {
```
Comment (pbenes) [inline src/components/visualizations/chart/highcharts/comboConfiguration.ts]: I think having the original `get(config, "primaryChartType", COLUMN);` was more safe. Default params are used only when the whole config would be missing. IChartConfig contains many other things, not just chart type config (such as colors, limits, etc.). So it would seem better not to assign DEFAULT_COMBO_CONFIG here because it will not be used when config will be given but will not contain primaryChartType and secondaryChartType.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3971/ : SUCCESS in 10m 06s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5026/ : SUCCESS in 7s

File: src/components/visualizations/chart/highcharts/comboConfiguration.ts [hunk 8]
```diff
@@ -1,29 +1,46 @@
...
 import { IChartConfig } from "../../../../interfaces/Config";
 import { VisualizationTypes } from "../../../../constants/visualizationTypes";
+import { isLineChart } from "../../utils/common";
 
 const { COLUMN, LINE } = VisualizationTypes;
 
 export function getDefaultChartType(config?: IChartConfig) {
-    const primaryType = get(config, "primaryChartType", COLUMN);
-    const secondaryType = get(config, "secondaryChartType", LINE);
+    const primaryChartType = get(config, "primaryChartType", COLUMN);
```
Comment (pbenes) [inline src/components/visualizations/chart/highcharts/comboConfiguration.ts]: what about adding a function (not tested)
```
function getDefaultComboTypes(config?: IChartConfig) {
    return {
        primaryType: get(config, "primaryChartType", COLUMN),
        secondaryType: get(config, "secondaryChartType", LINE)
    };
}

and then use on both places (here and here: https://github.com/gooddata/gooddata-react-components/pull/1025/files#diff-17f6209e1b24cc4a39babe31a1a4b51fR27) just 
`const { primaryChartType, secondaryChartType } = getDefaultComboTypes(config);`

It would be better to have "defaults" on only one place.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3973/ : FAILURE in 1m 56s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5029/ : SUCCESS in 10s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3974/ : SUCCESS in 5m 19s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5030/ : SUCCESS in 7s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3975/ : SUCCESS in 9m 04s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/5031/ : SUCCESS in 7s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2166/ : SUCCESS in 12m 34s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3976/ : SUCCESS in 8m 53s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1045/ : SUCCESS in 28s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1210/ : SUCCESS in 11m 30s
